From d0e2ce7a98cd606f0dabd512e129593ff6adb91a Mon Sep 17 00:00:00 2001
From: bananacakes <tahitiisland03@gmail.com>
Date: Sun, 16 Oct 2011 01:51:28 -0300
Subject: [PATCH] Add: Trackpad to wakeunlock

---
 drivers/input/misc/gpio_matrix.c |   12 +++++-------
 1 files changed, 5 insertions(+), 7 deletions(-)

diff --git a/drivers/input/misc/gpio_matrix.c b/drivers/input/misc/gpio_matrix.c
index f7379cd..2e02d5b 100644
--- a/drivers/input/misc/gpio_matrix.c
+++ b/drivers/input/misc/gpio_matrix.c
@@ -163,8 +163,9 @@ static void report_key(struct gpio_kp *kp, int key_index, int out, int in)
 		if (need_send_spec_key == pressed) {
 			curcial_oj_send_key(keycode, pressed);
 			need_send_spec_key = !pressed;
-			KEY_LOGI("%s: send OJ action key, pressed: %d\n",
-				__func__, pressed);
+			printk(KERN_INFO "%s: send OJ action key, pressed: %d\n",
+				__func__, need_send_spec_key);
+					input_report_key(kp->input_devs->dev[dev], keycode, pressed);
 		}
 	}
 #endif
@@ -392,11 +393,8 @@ int gpio_event_matrix_func(struct gpio_event_input_devs *input_devs,
 	if (func == GPIO_EVENT_FUNC_SUSPEND || func == GPIO_EVENT_FUNC_RESUME) {
 		/* TODO: disable scanning */
 		if (mi->detect_phone_status == 0) {
-			if (func == GPIO_EVENT_FUNC_SUSPEND)
-				irq_status = 0;
-			else
-				irq_status = 1;
-		} else {
+			return 0;
+			
 			phone_call_status = gpio_event_get_phone_call_status() & 0x01;
 			fm_radio_status = gpio_event_get_fm_radio_status() & 0x01;
 			KEY_LOGI("%s: mi->ninputs: %d, func&0x01 = %d, phone_call_status=%d, fm_radio_status=%d\n", __func__, mi->ninputs, func & 0x01, phone_call_status, fm_radio_status);
-- 
1.7.5.4

